src/Domain.LinnApps/Boms/BomChangeService.cs

```diff
@@ -345,13 +103,207 @@ public void DeleteAllFromBom(string bomName, int crfNumber, int changedBy)
+                                                     ? null : replacementSeq += 1,
+                DeleteChangeId = null,
+                DeleteReplaceSeq = null,
+                PcasLine = "N"
+            });
+        }
+
+        private void MaybeUpdateNode(BomTreeNode node, BomChange change)
+        {
+            var detail = this.bomDetailRepository.FindById(int.Parse(node.Id));
+            if (detail.Qty != node.Qty || detail.GenerateRequirement != node.Requirement)
+            {
+                if (detail.AddChange.DocumentNumber != change.DocumentNumber)
+                {
+                    throw new InvalidBomChangeException(
+                        "Can't directly update details added by a different CRF - Replace them instead!");
+                }
+
+                detail.Qty = node.Qty;
+                detail.GenerateRequirement = node.Requirement;
+            }
+        }
+
+        private void DeleteNode(BomTreeNode node, BomChange change)
+        {
+            var detail = this.bomDetailRepository.FindById(int.Parse(node.Id));
+            if (detail.DeleteChangeId.HasValue
+                && detail.DeleteChange?.DocumentNumber != change.DocumentNumber)
+            {
+                throw new InvalidBomChangeException(
+                    $"{node.Name} is already marked for deletion by another change request"
+                    + $" ({detail.DeleteChange?.DocumentNumber})");
+            }
+
+            if (detail.PcasLine == "Y")
+            {
+                throw new InvalidBomChangeException($"{node.Name} is a PCAS line - cannot delete here.");
+            }
+
+            if (detail.AddChange.DocumentNumber == change.DocumentNumber)
+            {
+                this.bomDetailRepository.Remove(detail);
+
+                if (detail.AddReplaceSeq.HasValue)
+                {
+                    var replacedDetail = this.bomDetailRepository.FindBy(
+                        x => x.DeleteChangeId == change.ChangeId
+                             && x.DeleteReplaceSeq == detail.AddReplaceSeq);
+                    replacedDetail.DeleteReplaceSeq = null;
+                    replacedDetail.DeleteChangeId = null;
+                }
+            }
+            else
+            {
+                detail.DeleteChangeId = change.ChangeId;
+            }
+        }
+
+        private void UndoNodeDeletion(BomTreeNode node, int changeRequestNumber)
+        {
+            // todo - is this correct? - if so add a unit test to enshrine it
+            if (node.DeleteChangeDocumentNumber.GetValueOrDefault() != changeRequestNumber)
+            {
+                throw new InvalidBomChangeException(
+                    $"cannot undo delete of {node.Name} - it was deleted on a different change request.");
+            }
+
+            node.DeleteChangeDocumentNumber = null;
+            node.ChangeState = "LIVE"; // todo - check this is right - add to test if so
```

Comment:
@andrewfraser73 any thoughts on this todo? My thinking is if you undo a delete node then that node needs to go back to being 'LIVE' - can you think of any situation where that won't be the case?